packages/gatsby/src/bootstrap/load-themes/index.js

```diff
@@ -5,25 +5,66 @@ const _ = require(`lodash`)
 const debug = require(`debug`)(`gatsby:load-themes`)
 const preferDefault = require(`../prefer-default`)
 const getConfigFile = require(`../get-config-file`)
+const reporter = require(`gatsby-cli/lib/reporter`)
 
 // get the gatsby-config file for a theme
-const resolveTheme = async themeSpec => {
+const resolveTheme = async (
+  themeSpec,
+  configFileThatDeclaredTheme,
+  isMainConfig = false
+) => {
   const themeName = themeSpec.resolve || themeSpec
   let themeDir
   try {
+    // theme is an node-resolvable module
     themeDir = path.dirname(require.resolve(themeName))
   } catch (e) {
-    // this can be local plugin, and require.resolve will throw
-    // in this case - let's return partial entry
-    return { themeName, themeSpec }
+    let panic = false
+    let pathToLocalTheme
+
+    // only try to look for local theme in main site
+    // local themes nested in other themes is potential source of problems:
+    // because those are not hosted by npm, there is potential for multiple
+    // local themes with same name that do different things and name being
+    // main identifier that gatsby uses right now, it's safer not to support it for now.
+    if (isMainConfig) {
+      // is a local plugin OR it doesn't exist
+      pathToLocalTheme = path.join(path.resolve(`.`), `plugins`, themeName)
+      try {
+        themeDir = path.dirname(require.resolve(pathToLocalTheme))
+      } catch (localErr) {
+        panic = true
+      }
+    } else {
+      panic = true
+    }
+
+    if (panic) {
```

Comment:
Nit: Moving this to finally might be a little more readable

Reply:
did some cleanup here (hopefully)